In [1]:
import numpy as np
from matplotlib import pyplot as plt
import psi4

Some notes:
At the FCI / cc-pVTZ level:

r_eq = 0.742427

# Monomer
At this geometry:

S0: -1.172336686493833 
The 20 most important determinants:

    *   1   -0.990975  (    0,    0)  1AX 
    *   2    0.055947  (    3,    3)  4AX 
    *   3    0.044436  (    1,    3)  2AA 4AB 
    *   4    0.044436  (    3,    1)  2AB 4AA 
    *   5    0.041860  (    5,    5)  6AX 
    *   6    0.041860  (    4,    4)  5AX 
    *   7    0.039689  (    1,    1)  2AX 
    *   8    0.038221  (    2,    2)  3AX 
    *   9    0.021207  (    3,   10)  4AA 11AB 
    *  10    0.021207  (   10,    3)  4AB 11AA 
    *  11    0.016679  (    2,   11)  3AA 12AB 
    *  12    0.016679  (   11,    2)  3AB 12AA 
    *  13    0.015879  (   11,   11)  12AX 
    *  14   -0.014814  (    2,    6)  3AA 7AB 
    *  15   -0.014814  (    6,    2)  3AB 7AA 
    *  16    0.014386  (   10,   10)  11AX 
    *  17    0.013272  (    1,   10)  2AA 11AB 
    *  18    0.013272  (   10,    1)  2AB 11AA 
    *  19    0.013103  (    6,    6)  7AX 
    *  20    0.010086  (   15,   15)  16AX 

S1:  -0.676605769856144

   The 20 most important determinants:

    *   1   -0.692247  (    1,    0)  1AB 2AA 
    *   2   -0.692247  (    0,    1)  1AA 2AB 
    *   3   -0.115069  (    1,    2)  2AA 3AB 
    *   4   -0.115069  (    2,    1)  2AB 3AA 
    *   5    0.057077  (    1,    6)  2AA 7AB 
    *   6    0.057077  (    6,    1)  2AB 7AA 
    *   7   -0.036264  (    3,    0)  1AB 4AA 
    *   8   -0.036264  (    0,    3)  1AA 4AB 
    *   9   -0.028216  (    1,   11)  2AA 12AB 
    *  10   -0.028216  (   11,    1)  2AB 12AA 
    *  11    0.022058  (    5,    7)  6AA 8AB 
    *  12    0.022058  (    7,    5)  6AB 8AA 
    *  13    0.022058  (    4,    8)  5AA 9AB 
    *  14    0.022058  (    8,    4)  5AB 9AA 
    *  15    0.021776  (    3,    6)  4AA 7AB 
    *  16    0.021776  (    6,    3)  4AB 7AA 
    *  17    0.014550  (    0,   10)  1AA 11AB 
    *  18    0.014550  (   10,    0)  1AB 11AA 
    *  19   -0.011322  (    1,   16)  2AA 17AB 
    *  20   -0.011322  (   16,    1)  2AB 17AA 

Transition Energy  = 0.495 a.u. or 13.49 eV at the FCI/cc-pVTZ level and transition dipole moment is (in atomic units)
 Dipole X            :          0.0000000            0.0000000            0.0000000
 Dipole Y            :         -0.0000000            0.0000000           -0.0000000
 Dipole Z            :          1.2103949            0.0000000            1.2103949
 Magnitude           :                                                    1.2103949

 ------------------------------------------------------------------------------------


 Using SA-CASSCF(2,2) for Roots 1 and 2 with equal weights, we get:

S0  -1.126445310417410
S1  -0.651987869958703
SA-CASSCF(2,2) Transition energy is 0.47445744045870697 a.u.
SA-CASSCF(2,2) Transition energy is 12.91065119529806 eV

TDM at the SA-CASSCF(2,2) level: 
 Dipole X            :          0.0000000            0.0000000            0.0000000
 Dipole Y            :         -0.0000000            0.0000000           -0.0000000
 Dipole Z            :          1.0931911            0.0000000            1.0931911
 Magnitude           :                                                    1.0931911


 # Dimer

 


 FCI Root 0 energy =    -2.344673372987338

 *   1    0.982032  (    0,    0)  1AX 2AX  (sigma bonding orbitals doubly occupied on each H2)
 *   2   -0.055442  (   15,   15)  1AX 7AX 

 FCI Root 1 energy =    -1.848942459468918

    *   1    0.679457  (    1,    0)  1AX 2AB 3AA (
    *   2    0.679457  (    0,    1)  1AX 2AA 3AB 
    *   3    0.112943  (    1,   10)  1AX 3AA 6AB 
    *   4    0.112943  (   10,    1)  1AX 3AB 6AA 

 FCI Root 2 energy =    -1.848942456287076

    *   1    0.679457  (    0,    4)  1AA 2AX 4AB 
    *   2    0.679457  (    4,    0)  1AB 2AX 4AA 
    *   3   -0.112943  (    7,    4)  2AX 4AB 5AA 
    *   4   -0.112943  (    4,    7)  2AX 4AA 5AB 

SCF Orbitals
    Doubly Occupied:                                                      

       1A     -0.593888     2A     -0.593888  

    Virtual:                                                              

       3A      0.166892     4A      0.166892     5A      0.305243  
       6A      0.305243     7A      0.642865     8A      0.642865  
       9A      0.692428    10A      0.692428    11A      0.692428  
      12A      0.692428    13A      1.095930    14A      1.095930  
      15A      1.119695    16A      1.119695    17A      1.119695  
      18A      1.119695    19A      1.528617    20A      1.528617  
      21A      2.337100    22A      2.337100    23A      2.408253  
      24A      2.408253    25A      3.175121    26A      3.175121  
      27A      3.175121    28A      3.175121    29A      3.411223  
      30A      3.411223    31A      3.411223    32A      3.411223  
      33A      3.674161    34A      3.674161    35A      3.895016  
      36A      3.895016    37A      3.895016    38A      3.895016  
      39A      4.082570    40A      4.082570    41A      4.082570  
      42A      4.082570    43A      4.112546    44A      4.112546  
      45A      4.112546    46A      4.112546    47A      4.651236  
      48A      4.651236    49A      5.722834    50A      5.722834  
      51A      5.722834    52A      5.722834    53A      5.798613  
      54A      5.798614    55A      6.957283    56A      6.957283  


      MCSCF Root 0 energy =    -2.263924527002497

         Active Space Natural occupation numbers:

         A   1.997010         A   1.997010         A   0.002990
         A   0.002990
         
    MCSCF Root 1 energy =    -1.776341149886159

   Active Space Natural occupation numbers:

         A   1.980252         A   1.016759         A   0.983241
         A   0.019748

     MCSCF Root 2 energy =    -1.776341146764290

   Active Space Natural occupation numbers:

         A   1.980252         A   1.016759         A   0.983241
         A   0.019748

In [13]:
print(F'Monomer States')
print(F'FCI Transition energy is {1.172336686493833 - 0.676605769856144} a.u.')
print(F'FCI Transition energy is {(1.172336686493833 - 0.676605769856144) * 27.2114} eV')

print(F'SA-CASSCF(2,2) Transition energy is {(1.126445310417410 - 0.651987869958703) * 1} a.u.')
print(F'SA-CASSCF(2,2) Transition energy is {(1.126445310417410 - 0.651987869958703) * 27.2114} eV')


print(F'Dimer State')
print(F'FCI Transition energy for |S0>|S0>->|S1>|S0> {2.344673372987338 - 1.848942459468918} a.u.')
print(F'FCI Transition energy for |S0>|S0>->|S0>|S1> {2.344673372987338 - 1.848942456287076} a.u.')
print(F'FCI Transition energy for |S0>|S0>->|S1>|S0> {(2.344673372987338 - 1.848942459468918) * 27.2114} eV')
print(F'FCI Transition energy for |S0>|S0>->|S0>|S1> {(2.344673372987338 - 1.848942456287076) * 27.2114} eV')

print(F'SA-CASSCF(4,4) Transition energy for |S0>|S0>->|S1>|S0> {2.263924527006536 - 1.776341149841512} a.u.')
print(F'SA-CASSCF(4,4) Transition energy for |S0>|S0>->|S0>|S1> {2.263924527006536 - 1.776341149840013} a.u.')
print(F'SA-CASSCF(4,4) Transition energy for |S0>|S0>->|S1>|S0> {(2.263924527006536 - 1.776341149841512) * 27.2114} eV')
print(F'SA-CASSCF(4,4) Transition energy for |S0>|S0>->|S0>|S1> {(2.263924527006536 - 1.776341149840013) * 27.2114} eV')

print(F'Size extensivity test')
print(F'FCI |S0>|S0> energy is 2x |S0> energy:')
print(np.isclose(1.172336686493833 * 2, 2.344673372987338))

print(F'Size extensivity test')
print(F'SA-CASSCF(4,4) |S0>|S0> energy is 2x  CASSCF(2,2) |S0> energy:')
print((1.126445310417410 * 2 - 2.263924527006536))




print(100*(1.0931 - 1.21) / 1.21 )

Monomer States
FCI Transition energy is 0.495730916637689 a.u.
FCI Transition energy is 13.48953226499481 eV
SA-CASSCF(2,2) Transition energy is 0.47445744045870697 a.u.
SA-CASSCF(2,2) Transition energy is 12.91065119529806 eV
Dimer State
FCI Transition energy for |S0>|S0>->|S1>|S0> 0.49573091351842 a.u.
FCI Transition energy for |S0>|S0>->|S0>|S1> 0.49573091670026215 a.u.
FCI Transition energy for |S0>|S0>->|S1>|S0> 13.489532180115136 eV
FCI Transition energy for |S0>|S0>->|S0>|S1> 13.489532266697514 eV
SA-CASSCF(4,4) Transition energy for |S0>|S0>->|S1>|S0> 0.487583377165024 a.u.
SA-CASSCF(4,4) Transition energy for |S0>|S0>->|S0>|S1> 0.487583377166523 a.u.
SA-CASSCF(4,4) Transition energy for |S0>|S0>->|S1>|S0> 13.267826309388335 eV
SA-CASSCF(4,4) Transition energy for |S0>|S0>->|S0>|S1> 13.267826309429125 eV
Size extensivity test
FCI |S0>|S0> energy is 2x |S0> energy:
True
Size extensivity test
SA-CASSCF(4,4) |S0>|S0> energy is 2x  CASSCF(2,2) |S0> energy:
-0.011033906171716179
-9.

In [3]:
# optimized at the FCI/cc-pVTZ level
mol_str = """
H
H 1 0.742427
symmetry c1
"""

monomer_cart = """
H 0.0000000000   0.0000000000  -0.3713013354
H 0.0000000000   0.0000000000   0.3713013354
symmetry c1
"""

dimer_cart = """
H 0.0000000000   0.0000000000  -0.3713013354
H 0.0000000000   0.0000000000   0.3713013354
H 0.0000000000   0.0000000000   999.62869866
H 0.0000000000   0.0000000000   1000.37130134
symmetry c1
"""

trimer_cart = """
H 0.0000000000   0.0000000000  -0.3713013354
H 0.0000000000   0.0000000000   0.3713013354
H 0.0000000000   0.0000000000   999.62869866
H 0.0000000000   0.0000000000   1000.37130134
H 0.0000000000   0.0000000000   -999.62869866
H 0.0000000000   0.0000000000   -1000.37130134
symmetry c1
"""


trimer_dict = {
    "basis": "cc-pVTZ",
    "scf_type": "pk",
    "e_convergence": 1e-10,
    "d_convergence": 1e-10,
    'num_roots' : 1,
    "frozen_docc" : [0],
    "active" : [6],
    "mcscf_e_convergence" : 1e-9,
    "mcscf_r_convergence" : 1e-6
}

dimer_dict = {
    "basis": "cc-pVTZ",
    "scf_type": "pk",
    "e_convergence": 1e-10,
    "d_convergence": 1e-10,
    'num_roots' : 1,
    "frozen_docc" : [0],
    "active" : [4],
    "mcscf_e_convergence" : 1e-9,
    "mcscf_r_convergence" : 1e-6
}


monomer_dict = {
    "basis": "cc-pVTZ",
    "scf_type": "pk",
    "e_convergence": 1e-10,
    "d_convergence": 1e-10,
    'num_roots' : 1,
    "frozen_docc" : [0],
    "active" : [2],
    "mcscf_e_convergence" : 1e-9,
    "mcscf_r_convergence" : 1e-6
}


species = "trimer"


#fci_energy, wfn = psi4.energy('fci',ref_wfn=wfn, return_wfn=True)
#casscf_energy_3_6, wfn = psi4.energy('casscf', ref_wfn=wfn, return_wfn=True)


In [4]:
if species=="monomer":
    monomer_mol = psi4.geometry(monomer_cart)
    psi4.set_options(monomer_dict)
    monomer_e, monomer_wfn = psi4.energy('SCF', molecule=monomer_mol, return_wfn=True)
    cas_monomer_e = psi4.energy('casscf', molecule=monomer_mol, ref_wfn=monomer_wfn, return_wfn=False)

elif species=="dimer":
    dimer_mol = psi4.geometry(dimer_cart)
    psi4.set_options(dimer_dict)
    dimer_e, dimer_wfn = psi4.energy('SCF', molecule=dimer_mol, return_wfn=True)
    cas_dimer_e = psi4.energy('casscf', molecule=dimer_mol, ref_wfn=dimer_wfn, return_wfn=False)

elif species=="trimer":
    trimer_mol = psi4.geometry(trimer_cart)
    psi4.set_options(trimer_dict)
    trimer_e, trimer_wfn = psi4.energy('SCF', molecule=trimer_mol, return_wfn=True)
    cas_trimer_e = psi4.energy('casscf', molecule=trimer_mol, ref_wfn=trimer_wfn, return_wfn=False)


Scratch directory: /tmp/

*** tstart() called on Jessicas-iMac.local
*** at Wed Sep 18 12:35:28 2024

   => Loading Basis Set <=

    Name: CC-PVTZ
    Role: ORBITAL
    Keyword: BASIS
    atoms 1-6 entry H          line    23 file /opt/anaconda3/envs/psi4env/share/psi4/basis/cc-pvtz.gbs 


         ---------------------------------------------------------
                                   SCF
               by Justin Turney, Rob Parrish, Andy Simmonett
                          and Daniel G. A. Smith
                              RHF Reference
                        1 Threads,    500 MiB Core
         ---------------------------------------------------------

  ==> Geometry <==

    Molecular point group: c1
    Full point group: D_inf_h

    Geometry (in Angstrom), charge = 0, multiplicity = 1:

       Center              X                  Y                   Z               Mass       
    ------------   -----------------  -----------------  -----------------  -----------------


   @MCSCF  5:     -3.454283561157   -2.9269e-04  3.21e-04  3.72e-13    3    1  TS, DIIS
   @MCSCF  6:     -3.454347807351   -6.4246e-05  6.32e-05  2.52e-13    3    1  TS, DIIS
   @MCSCF  7:     -3.454348123566   -3.1621e-07  2.10e-05  4.06e-14    3    1  TS, DIIS
   @MCSCF  8:     -3.454348227420   -1.0385e-07  4.14e-06  4.01e-14    3    1  TS, DIIS
   @MCSCF  9:     -3.454348231492   -4.0715e-09  5.53e-08  3.03e-14    3    1  TS, DIIS
   @MCSCF 10:     -3.454348231495   -2.6610e-12  1.13e-08  3.33e-14    3    1  TS, DIIS

          @MCSCF has converged!

   @MCSCF Final Energy:   -3.454348231494644

   Computing CI Semicanonical Orbitals

   ==> Energetics <==

    SCF energy =           -3.398827744509924
    Total MCSCF energy =   -3.454348231494644

   ==> MCSCF root 0 information <==

    MCSCF Root 0 energy =    -3.454348231494647

   Active Space Natural occupation numbers:

         A   1.975817         A   1.975817         A   1.975817
         A   0.024183         A   0.02418

In [5]:
print(trimer_e)
print(cas_trimer_e)

-3.3988277445099238
-3.454348231494647


In [6]:
0.01 * np.sqrt(2)

0.014142135623730952

In [7]:
monomer_e = -1.1329425815734104
cas_monomer_e = -1.1514494103995814

dimer_e = -2.265885163041723
cas_dimer_e = -2.3028988209471546

trimer_e = -3.3988277445099246
cas_trimer_e = -3.4543482314946363

print(F"SCF Monomer Energy is {monomer_e:12.10f}")
print(F"SCF Dimer Energy is {dimer_e:12.10f}")
print(F"2xSCF Monomer Energy is {2 * monomer_e:12.10f}")
print(F"SCF Trimer Energy is {trimer_e:12.10f}")
print(F"3xSCF Monomer Energy is {3 * monomer_e:12.10f}")

scf_sc_error = trimer_e - 3 * monomer_e
casscf_sc_error = cas_trimer_e - 3 * cas_monomer_e

print(F"SCF Size Extensivity Error: {scf_sc_error:12.10e}")
print(F"SCF Size Extensivity Error: {casscf_sc_error:12.10e}")

#print(F"CASSCF Monomer Energy is {cas_monomer_e:12.10f}")
#print(F"CASSCF Dimer Energy is {cas_dimer_e:12.10f}")
#print(F"2xCASSCF Monomer Energy is {2 * cas_monomer_e:12.10f}")
#print(F"CASSCF Trimer Energy is {cas_trimer_e:12.10f}")
#print(F"3xCASSCF Monomer Energy is {3 * cas_monomer_e:12.10f}")


SCF Monomer Energy is -1.1329425816
SCF Dimer Energy is -2.2658851630
2xSCF Monomer Energy is -2.2658851631
SCF Trimer Energy is -3.3988277445
3xSCF Monomer Energy is -3.3988277447
SCF Size Extensivity Error: 2.1030643893e-10
SCF Size Extensivity Error: -2.9589219963e-10


In [8]:
print(monomer_e, 2*monomer_e, 3*monomer_e)
print(dimer_e)
print(trimer_e)

-1.1329425815734104 -2.265885163146821 -3.398827744720231
-2.265885163041723
-3.3988277445099246


# equilibrium geometry of H2 at the FCI/cc-pVTZ level
r_eq = 0.742427
fci_energy = -1.172336661363193

# com coordinates
H_1 = np.array([0.0000000000,   0.0000000000,  -0.3713013354])
H_2 = np.array([0.0000000000,   0.0000000000,   0.3713013354])

FCI Root 0 energy =    -1.172336661363193

CASSCF(2,8)
   The 20 most important determinants:

    *   1    0.990978  (    0,    0)  1AX 
    *   2   -0.055938  (    3,    3)  4AX 
    *   3   -0.044420  (    1,    3)  2AA 4AB 
    *   4   -0.044420  (    3,    1)  2AB 4AA 
    *   5   -0.041858  (    4,    4)  5AX 
    *   6   -0.041858  (    5,    5)  6AX 
    *   7   -0.039666  (    1,    1)  2AX 
    *   8   -0.038219  (    2,    2)  3AX 
    *   9   -0.021208  (    3,   10)  4AA 11AB 
    *  10   -0.021208  (   10,    3)  4AB 11AA 
    *  11   -0.016682  (    2,   11)  3AA 12AB 
    *  12   -0.016682  (   11,    2)  3AB 12AA 
    *  13   -0.015883  (   11,   11)  12AX 
    *  14    0.014814  (    2,    6)  3AA 7AB 
    *  15    0.014814  (    6,    2)  3AB 7AA 
    *  16   -0.014389  (   10,   10)  11AX 
    *  17   -0.013271  (    1,   10)  2AA 11AB 
    *  18   -0.013271  (   10,    1)  2AB 11AA 
    *  19   -0.013101  (    6,    6)  7AX 
    *  20   -0.010089  (   14,   14)  15AX 

    MCSCF Root 0 energy =    -1.169422343006941

   Active Space Natural occupation numbers:

         A   1.963897         A   0.021110         A   0.006067
         A   0.004417         A   0.004345         A   0.000150
         A   0.000012         A   0.000002

CASSCF(2,2)

    MCSCF Root 0 energy =    -1.151446562735611

   Active Space Natural occupation numbers:

         A   1.975827         A   0.024173

        A   1.975813         A   1.975813         A   0.024187
         A   0.024187

CASSCF(4,4) for 2 H2 large separation

  MCSCF Root 0 energy =    -1.158678336763447

  Active Space Natural occupation numbers: 
         A   1.970437         A   0.023040         A   0.006521
         A   0.000002


CASSCF(6,6) for 3 H2 large separation
         A   1.975817         A   1.975817         A   1.975817
         A   0.024183         A   0.024183         A   0.024183

   SCF energy =           -3.398827744509925
   MCSCF Root 0 energy =    -3.454348231491987       

In [10]:
# equilibrium geometry of H2 at the FCI/cc-pVTZ level
r_eq = 0.742427
fci_energy = -1.172336661363193

fci_energy_2H2 = -2.344673372984059

casscf_6_6_3H2 = -3.454348231491987
casscf_4_4_2H2 = -2.302898818231016
casscf_2_2 = -1.151446562735611
casscf_2_4 = -1.158678336763446
casscf_2_5 = -1.163930435072928
casscf_2_6 = -1.169020481381448
print(fci_energy-casscf_2_4)
print(fci_energy-casscf_2_5)
print(fci_energy-casscf_2_6)
print(2*fci_energy-fci_energy_2H2)

print(casscf_4_4_2H2 - 2 * casscf_2_2)
print(casscf_6_6_3H2 - 3 * casscf_2_2)

# com coordinates
H_1 = np.array([0.0000000000,   0.0000000000,  -0.3713013354])
H_2 = np.array([0.0000000000,   0.0000000000,   0.3713013354])

disp = np.array([0, 0, 12000])
H_3 = H_1 - disp
H_4 = H_2 - disp

print(H_1 + disp)
print(H_2 + disp)
print(H_3)
print(H_4)



-0.01365832459974703
-0.008406226290264884
-0.0033161799817449378
5.0257673134268543e-08
-5.692759794229119e-06
-8.543285153983504e-06
[    0.             0.         11999.62869866]
[    0.             0.         12000.37130134]
[     0.              0.         -12000.37130134]
[     0.              0.         -11999.62869866]
